### PR TITLE
DSDEGP-1703 Support downloading files to another Google bucket

### PR DESCRIPTION
**Pull request by vvicario**
### Description
Cancel is being handled in other story. This branch needs npm install.

----

### Checklist

- [] **Submitter**: Include the JIRA issue number in the PR description
- [] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [] **Submitter**: Ensure that you have added or updated tests
- [] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=189431877).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

### CODE REVIEW

File: src/app/file-manager/file-explorer/file-explorer.component.ts [hunk 1]
```diff
@@ -67,7 +77,7 @@ export class FileExplorerComponent implements OnInit, AfterViewInit {
         this.zone.run(() => {
           if (data === 100) {
             this.downloadInProgress = false;
-          } else {
+          } else if (this.router.routerState.snapshot.url === '/file-download') {
```
Comment (pshapiro4broad) [inline src/app/file-manager/file-explorer/file-explorer.component.ts]: Why is this necessary? Is it OK that both branches of the `if` `else` are false then `downloadInProgress` won't be updated at all?

File: src/app/file-manager/file-explorer/file-explorer.component.ts [hunk 2]
```diff
@@ -39,20 +45,24 @@ export class FileExplorerComponent implements OnInit, AfterViewInit {
   headerItem: Item;
   rootItem: Item;
   downloadInProgress = false;
+  exportInProgress = false;
```
Comment (pshapiro4broad) [inline src/app/file-manager/file-explorer/file-explorer.component.ts]: field appears to be unused

File: src/app/file-manager/file-explorer/file-explorer.component.ts [hunk 3]
```diff
@@ -338,4 +352,28 @@ export class FileExplorerComponent implements OnInit, AfterViewInit {
...
+      width: '500px',
+      disableClose: true,
+      data: items
+    });
+    dialogRef.afterClosed().subscribe(result => {
+      if (result.cancel === undefined) {
+        this.transferablesGridComponent.startExportToGCP(result.preserveStructure);
+      } else {
+        const itemsToRemove = new FilesDatabase(this.store).data
+        .filter(item => item.type === Type.EXPORT_GCP && item.status === ItemStatus.PENDING);
```
Comment (pshapiro4broad) [inline src/app/file-manager/file-explorer/file-explorer.component.ts]: `.filter` should be indented here

File: src/app/file-manager/services/limit-transferables.service.ts [hunk 4]
```diff
@@ -55,11 +55,22 @@ export class LimitTransferablesService {
...
-    } else {
+    } else if (type === Type.UPLOAD) {
       this.gcsService.uploadFiles(localStorage.getItem('uploadBucket'), maxFiles);
     }
   }
 
+
+  public exportItems(files: Item[]): void {
+    let maxFiles = [];
+    if (files.length > environment.LIMIT_EXPORTABLES) {
```
Comment (pshapiro4broad) [inline src/app/file-manager/services/limit-transferables.service.ts]: Is this `if` necessary? I would expect that `splice` would remove all items if there were less than `LIMIT_EXPORTABLES` items left.

File: src/app/file-manager/file-export-modal/file-export-modal.component.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+import { FilesService } from '@app/file-manager/services/files.service';
+import { OnInit } from '@angular/core/src/metadata/lifecycle_hooks';
+import { Type } from '@app/file-manager/models/type';
+import { ItemStatus } from '@app/file-manager/models/item-status';
+import { Item } from '@app/file-manager/models/item';
+import * as Transferables from '../actions/transferables.actions';
+import { GcsService } from '@app/file-manager/services/gcs.service';
+import { Store } from '@ngrx/store';
+import { AppState } from '../dbstate/app-state';
+import { MessageService } from 'primeng/components/common/messageservice';
```
Comment (pshapiro4broad) [inline src/app/file-manager/file-export-modal/file-export-modal.component.ts]: `MessageService` appears to be unused

File: src/app/file-manager/file-export-modal/file-export-modal.component.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+@Component({
+  selector: 'app-file-export-modal',
+  templateUrl: './file-export-modal.component.html',
+  styleUrls: ['./file-export-modal.component.scss']
+})
+export class FileExportModalComponent implements OnInit {
+
+  exportForm: FormGroup;
+  preserveStructure = true;
+  disableExport = true;
```
Comment (pshapiro4broad) [inline src/app/file-manager/file-export-modal/file-export-modal.component.ts]: `disableExport` appears to be unused

File: src/app/file-manager/file-export-modal/file-export-modal.component.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    public dialogRef: MatDialogRef<FileExportModalComponent>,
+    @Inject(MAT_DIALOG_DATA) public data: any,
+    private router: Router,
+    private formBuilder: FormBuilder,
+    private store: Store<AppState>,
+    private gcsService: GcsService,
+    private transferablesGridComponent: TransferablesGridComponent,
+    private filesService: FilesService,
+    private bucketService: BucketService,
+    private zone: NgZone
```
Comment (pshapiro4broad) [inline src/app/file-manager/file-export-modal/file-export-modal.component.ts]: `zone` appears to be unused

File: src/app/file-manager/file-export-modal/file-export-modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+import { TransferablesGridComponent } from '@app/file-manager/transferables-grid/transferables-grid.component';
+import { BucketService } from '@app/file-manager/services/bucket.service';
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+
+@Component({
+  selector: 'app-file-export-modal',
+  templateUrl: './file-export-modal.component.html',
+  styleUrls: ['./file-export-modal.component.scss']
+})
+export class FileExportModalComponent implements OnInit {
```
Comment (pshapiro4broad) [inline src/app/file-manager/file-export-modal/file-export-modal.component.ts:20]: It looks like this duplicates a lot of the code in `FileDownloadModalComponent`. It would be better if there was a common base class that contained the shared code, to avoid this duplication.

File: src/assets/export_to_cloud_dark.svg [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (pshapiro4broad) [inline src/assets/export_to_cloud_dark.svg]: Maybe instead of `dark` a better name would be `selected`? When I see 'dark' it makes me think of the 'dark' themes that some websites have.

File: src/assets/export_to_cloud.svg [hunk 10]
```diff
@@ -0,0 +1 @@
+<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><defs><style>.cls-1{fill:#fff;}</style></defs><title>export_to_cloud</title><path class="cls-1" d="M18.81,8.73A7.44,7.44,0,0,0,5.32,8.12,6,6,0,0,0,6,20H18.88a5,5,0,0,0,4.37-2.65A5.27,5.27,0,0,0,24,14.63,6,6,0,0,0,18.81,8.73Zm2.69,7.82A3.11,3.11,0,0,1,18.88,18H6a4,4,0,0,1-3.78-2.8h0a4,4,0,0,1,2.71-5A3.94,3.94,0,0,1,6,10.06h.71a5.46,5.46,0,0,1,10.62.47l0,.12H18a4,4,0,0,1,4,4A3.93,3.93,0,0,1,21.5,16.55Z"/><polygon class="cls-1" points="12.9 9.79 16 12.89 12.9 15.99 12.9 14.13 8 14.13 8 11.65 12.9 11.65 12.9 9.79"/></svg>
```
Comment (pshapiro4broad) [inline src/assets/export_to_cloud.svg]: Not sure why this is not wrapped but the `dark` one is. Either way is OK but it would be nice to be consistent.

File: src/app/file-manager/transferables-grid/transferables-grid.component.ts [hunk 11]
```diff
@@ -15,6 +15,11 @@ import { FilesDatabase } from '../dbstate/files-database';
 import { LimitTransferablesService } from '../services/limit-transferables.service';
 import { Type } from '@app/file-manager/models/type';
 import { ItemStatus } from '@app/file-manager/models/item-status';
+import { NgxSpinnerComponent } from 'ngx-spinner';
+import { NgxSpinnerService } from 'ngx-spinner';
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+import { modelGroupProvider } from '@angular/forms/src/directives/ng_model_group';
```
Comment (pshapiro4broad) [inline src/app/file-manager/transferables-grid/transferables-grid.component.ts]: Some of these imports don't appear to be used: `NgxSpinnerComponent`, `BehaviorSubject`, `modelGroupProvider`.

**Review by pshapiro4broad — COMMENTED** [review]
Just a few minor comments.

File: src/assets/export_to_cloud_dark.svg (hunk 9 above)
Comment (nlopezzalba) [reply src/assets/export_to_cloud_dark.svg]: I called it like that to differentiate it from the white colored version, maybe I should name them export_to_cloud_white and export_to_cloud_grey? BTW this version is only being used in the status table for the moment

File: src/assets/export_to_cloud.svg (hunk 10 above)
Comment (nlopezzalba) [reply src/assets/export_to_cloud.svg]: You're right, already fixed!

**Comment by pshapiro4broad** [comment]
The name 'dark' is fine if it's just a darker version of the image, I had thought that the code was using it to show the selected state of the icon.

**Comment by vvicario** [comment]
@pshapiro4broad  I worked on your feedback, please let me know if you have any other comment. Thanks!

File: main.js [hunk 12]
```diff
@@ -59,7 +59,7 @@ app.on('ready', function () {
 
   // Show dev tools
   // Remove this line before distributing
-  // win.webContents.openDevTools();
+   win.webContents.openDevTools();
```
Comment (pshapiro4broad) [inline main.js]: don't forget to remove before merging

**Review by pshapiro4broad — APPROVED** [review]
need to remove dev window, looks OK otherwise

File: main.js (hunk 12 above)
Comment (vvicario) [reply main.js]: (:+1: )
